### DISCUSSION TITLE
Add sweep_frequency service to RemoteEntity

### DISCUSSION THREAD

**Discussion by felipediel** [home-assistant/architecture#1076] [Entity Models] [answered]
## Context

In the process of scanning commands from RF remotes, it's essential to first scan the frequency. Currently, both tasks are managed by the same service — learn_command — which presents usability challenges for learning RF commands due to the following reasons:

- Scanning the frequency is the most time-consuming step in the process, taking 5~10 seconds. Currently, users are required to scan the frequency repeatedly for each learned command, instead of performing it just once initially.
- The user should interact with the UI, not with the service. Present implementations involve users staying alert to notifications indicating when to initiate a long click (to sweep frequency) followed by a short click (to learn the command), which is not very user-friendly.

## Proposal
A preferable approach involves separating these tasks into distinct services — one dedicated to identifying the frequency and another to learn the commands themselves. This separation of concerns offers several benefits, including:

- Users can sweep the frequency only once, separately.
- Components can leverage this to create better UIs and voice services. Learning can be divided into steps, showing appropriate messages that guide the user through the process eg. "Press and hold any button on the remote for 5 seconds" then "Press the power button" etc
- Better exception handling, informative error messages at each step.
- Users can create intents like "Alexa, what is the frequency of this remote?"

## Impact
Only two integrations have implemented the learn_command service so far:

- broadink
- mysensors

The only one that supports learning RF commands is Broadlink, so it will be the only one impacted by this change.

## Proof of Concept
https://github.com/home-assistant/core/pull/115785
https://github.com/home-assistant/core/pull/115791

**Comment by clockbrain** [comment]
I very much support this proposal. I had much difficulty in scanning my RF remote ceiling fan controllers though the HA UI remote learn service. Main problem was that my remotes are of the type where a long press (to keep transmitting for the frequency sweep) puts the remote into a different mode and the subsequent short press gets recorded as a different command to the one I was trying to record. I eventually gave up on using the HA UI learn approach and instead used the fixed frequency patch of the python-broadlink project to scan my remotes which worked much more reliably.

One thing I'd like to add is that I would like to see HA UI have the option of being able to manually specify the frequency instead of calling the sweep service. I know all my remotes are 433.92MHz so I see no need to sweep unless that is just a once-off and the frequency can be saved for many subsequent learn service calls.

**Comment by kobemtl** [comment]
You are the man! Just got the new Broadlink and the HA integration is way slower than the original app. This would be the solution! Thank you!

**Comment by CitizenRacer** [comment]
This would be so incredibly useful. Please add it!

**Comment by flequi19** [comment]
yes please!

**Comment by emontnemery** [comment]
Instead of the new service, would it be more natural to make the sweep frequency a parameter to the `learn_command` service?

**Reply by emontnemery** [reply]
@felipediel Could you comment on this suggestion please?

**Reply by felipediel** [reply]
Hi @emontnemery, sure, I will do it later today.

**Reply by emontnemery** [reply]
Another option would be a custom service to the broadlink integration

**Reply by felipediel** [reply]
Thank you for your feedback.  I apologize for the delay in getting back to you. Here are my thoughts on the options you proposed and why I believe the separation into distinct services, as initially suggested, remains the preferable approach:

1. **Sweep Frequency as a Parameter to learn_command**

**Separation of Concerns**: Treating frequency scanning and command learning as separate operations allows for clearer, more focused user interaction, and better maintainability of the codebase. Each task can be managed independently, providing more specific error handling and user instructions. Combining these tasks into a single service would require interacting with the same service using different parameters and returning different types of responses depending on the functionality. This can lead to a less clean design, where a single service is tasked with multiple, distinct responsibilities. A more modular approach not only improves the clarity and organization of the codebase, but also facilitates testing, documenting and interacting with these services.

2. **Custom Service in Broadlink Integration**

**Interface Segregation**: Implementing a dedicated service for frequency scanning in the Broadlink integration could limit the reusability of this functionality. By creating a separate, more generic frequency scanning service interface, we allow for broader use cases beyond just the Broadlink integration. Looking ahead, there is potential for developing a unified frontend for interacting with various universal remotes. As part of this future vision, we would need to model a common interface for these integrations, all of which require RF frequency scanning as a fundamental part of their operation.

**Reply by emontnemery** [reply]
Thanks @felipediel for the feedback. I don't understand why it's more user friendly to set the sweep frequency in a separate service.
You write "each task can be managed independently", however these are not separate tasks, it's a single task to learn a new command?

The second point is not valid because only broadlink currently can benefit from configurable `sweep_frequency`. To modify the entity model we require more than a single integration to benefit from the change.

**Reply by clockbrain** [reply]
I'll avoid commenting on the internals of HA, but from my perspective splitting this into two services gives a cleaner service interaction. Note that sweeping for a frequency takes many seconds and requires the user holding or repeatedly pressing a button on the remote which in my case puts the remote into a different mode and changes the function of the button I am trying to learn.

Here is what a sequence might look like as a single composite sweep-learn service with parameters:
- sweep-learn (sweep=true, freq=null) -> press & hold fan speed 1 button 5 seconds for sweep, press again for learn and hope mode hasn't changed and it still transmits code for fan speed 1 ->returns 433Mhz & bitstream for fan speed 1
- sweep-learn (sweep=false, freq=433Mhz) -> press fan speed 2 button -> returns bitstream for fan speed 2
- sweep-learn (sweep=false, freq=433Mhz) -> press fan speed 3 button -> returns bitstream for fan speed 3
- sweep-learn (sweep=false, freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light on
- sweep-learn (sweep=false, freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light off

Or, if the frequency is already known by the user and provided by the UI then no need to sweep:
- sweep-learn (sweep=false, freq=433Mhz) -> press fan speed 1 button -> returns bitstream for fan speed 1
- sweep-learn (sweep=false, freq=433Mhz) -> press fan speed 2 button -> returns bitstream for fan speed 2
- sweep-learn (sweep=false, freq=433Mhz) -> press fan speed 3 button -> returns bitstream for fan speed 3
- sweep-learn (sweep=false, freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light on
- sweep-learn (sweep=false, freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light off


Here is what a sequence might look like if there are separate sweep and learn services
- sweep () -> press and hold 5 seconds any button -> returns 433Mhz
- learn (freq=433Mhz) -> press fan speed 1 button -> returns bitstream for fan speed 1
- learn (freq=433Mhz) -> press fan speed 2 button -> returns bitstream for fan speed 2
- learn (freq=433Mhz) -> press fan speed 3 button -> returns bitstream for fan speed 3
- learn (freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light on
- learn (freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light off

Or, if the frequency is already known by the user and provided by the UI then no need to sweep:
- learn (freq=433Mhz) -> press fan speed 1 button -> returns bitstream for fan speed 1
- learn (freq=433Mhz) -> press fan speed 2 button -> returns bitstream for fan speed 2
- learn (freq=433Mhz) -> press fan speed 3 button -> returns bitstream for fan speed 3
- learn (freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light on
- learn (freq=433Mhz) -> press fan light toggle button -> returns bitstream for fan light off

**Reply by emontnemery** [reply]
@clockbrain The way I understand the proposal from @felipediel is different, I think what he proposes is this:
```
remote.sweep_frequency({"entity_id": "remote.living_room", "frequency":"433MHz"
remote.learn_command({"entity_id": "remote.living_room", "command": "fan_speed_1") -> press fan speed 1 button -> learns bitstream for fan speed 1, only listens at 433Mhz because of the previous `remote.sweep_frequency` service call
# One week passes
remote.learn_command({"entity_id": "remote.living_room", "command": "something") -> press a button on some other remote which is not the 433 MHz remote for the fan, learning fails because only 433MHz was tried
```

My concern is that a service call to set the frequency at some previous time affects the behavior of future calls to `remote.learn_command`

Edit: I misunderstood, it's clear now. Thanks for explaining @felipediel 👍

**Reply by felipediel** [reply]
The proposal includes adding a frequency parameter to the `learn_command` service, enabling it to be called with the correct frequency once it’s determined, eliminating the need for re-scanning. While Broadlink will serve as the pilot for this feature, it would benefit any RF universal remote integration, as they all follow a similar process: first, we sweep the frequency, and then we bind to it to capture the command.

**Reply by felipediel** [reply]
We can also introduce a feature flag, SUPPORT_SWEEP_FREQ, to ensure this change doesn't affect integrations that don't support this.

**Reply by emontnemery** [reply, marked as answer]
There are only two integrations which implement `remote.learn_command`: `broadlink` and `mysensors`. Based on the code, `mysensors` seems to be about learning IR buttons so setting a frequency is irrelevant for it.

Unless there are concrete examples of core integrations which could make use of this proposal, we should add two custom services to broadlink instead of changing the remote entity model:
- `broadlink.sweep_frequency` (I think the name `sweep_frequency` is unclear, I'd suggest something like `broadlink.detect_remote_frequency` instead)
- `broadlink.fast_learn_command` (A custom `learn_command` which accepts the `frequency` parameter, mabye the name `fast_learn_command` is not good)

If there are more integrations which could benefit in the future, we can revisit this discussion then; the proposal is good but I just don't see the need.

Also, thanks for the patience with explaining the proposal @felipediel, I didn't realize the `sweep_frequency` service communicates with the user via persistent notifications.

Communicating via persistent notifications is not something we'd allow, instead the service should return the frequency on which it detected the remote in a response: <https://developers.home-assistant.io/docs/dev_101_services#response-data>

**Reply by felipediel** [reply]
Could we at least consider adding a `frequency` parameter to the `learn_command` service? This would eliminate the need to rely on persistent notifications for communication.

Since the service was originally designed to support RF, this isn’t introducing Broadlink-specific functionality, but rather enhancing the interface to allow better RF integration without the cumbersome notification-based approach.

**Reply by emontnemery** [reply]
@felipediel We can consider that, if there are multiple integrations making use of it. We won't extend the functionality of `remote.learn_command` when only `broadlink` will benefit.

Could you share your concerns with implementing `broadlink` specific services?

**Reply by felipediel** [reply]
The `remote.learn_command` service was meant to create a unified interface for all integrations, enabling a unified frontend in the feature. Adding a Broadlink-specific service feels like a step backward.

**Reply by Comisionado8** [reply]
Any updates on this feature - Many people already know the frequency and having a sweep function not only make it more difficult to learn codes - with remote where holding or clicking puts the device into a difference mode it makes it impossible.

**Reply by landolfi-us** [reply]
Regardless of how this is implemented, it seems an essential feature. RF learning seems to be hit and miss as it is, and having to wait for a frequency scan when the frequency is already known upfront is a kludgy user experience and makes a time-consuming process even more time-consuming and prone to failure.

**Reply by christopherlam** [reply]
Wish to add support to the idea of `learn_frequency` type command, and the frequency can be offered to `learn_command` as a parameter. My RF remote sends only 1 impulse apparently therefore it's impossible to learn its RF Signal within HA.

**Reply by garymathews** [reply]
Since Broadlink is the only device that supports the `rf` option in the `learn_command`, it makes sense to include a `frequency` parameter. Without it the RF learning functionality of the Broadlink device is broken for single pulse devices (most RF remotes). It also doesn't make sense to sweep the frequency to learn each command.

Regarding the entity model, in its current state - assuming all future RF devices support a frequency sweep is not unified and future proof. A `frequency` parameter should be included that is optional - future and current devices that support frequency sweeps can fallback to the sweep if a frequency is not specified. This does not break any current implementations and allows for future devices to be included that support this parameter.

For anyone wanting this asap, you can check out this PR: https://github.com/home-assistant/core/pull/150317

**Reply by felipediel** [reply]
I venture that all RF blasters work the same way: first scanning the frequency and then capturing commands. I don’t see why Broadlink can’t lead this functionality, as it did with the learn command, delete command, and command type features. Those wouldn’t exist if we strictly followed the "more than one beneficiary" rule, and now many integrations benefit.

**Reply by inventor7777** [reply]
I'd like to offer my support for this function as well. It makes it very hard to use the Broadlink I specifically got for Home Assistant.

**Reply by make-all** [reply]
Although not part of HA core integrations, the possibility to pass a frequency to remote.learn_command has been requested for tuya_local, as Tuya RF remotes that support more than just 433MHz do not support scanning and need the frequency explicitly specified. I would much prefer to follow a standard for this than implement custom services, and if the Tuya built in integration gets support for remote entities one day it would also benefit from the same.

**Comment by arrrroooooo** [comment]
As a new user to Home Assistant, this functionality introduced me to the Home Assistant ecosystem and dev team with a significantly negative view. I fought for hours trying to get a simple RF command learned for my pre-installed curtains and eventually had to give up. Double slap to the face to find out it was due to a bug (this is not a feature request, it's broken functionality plain and simple) in one of the most recommended RF transceivers in the community. Triple slap: we're not blocked on finding a fix, but on philosophical quibbles. Egad.

**Reply by felipediel** [reply]
This is not a bug. Learning RF commands is entirely feasible, and instructions are available in the documentation. The process is still tricky because Home Assistant’s remote component architecture isn’t designed for it, so we need to handle multiple things within a single service call. Adding a `frequency` parameter to the `learn_command` service would solve this, but the decision is not up to me.

I’ve already written the code, provided suggestions, and linked everything here. Perhaps my mistake was proposing too many changes in a single discussion.

If anyone wants to continue, I suggest opening a new discussion focused solely on adding the frequency parameter. This thread has run its course, and further comments here are unlikely to be productive.
